tests/test_violations/test_docs.py

```diff
@@ -24,6 +24,15 @@ def test_violation_name(all_violations):
         assert class_name.endswith('Violation'), class_name
 
 
+def test_violations_with_previous_codes_have_versionchanged(all_violations):
+    """Ensures that all violations with the previous codes attribute
+    also have the same or greater number of ..versionchanged markers"""
+    for violation in all_violations:
+        if hasattr(violation, 'previous_codes'):
```

Comment:
We don't use `hasattr` in our code. We use `getattr`. In this case it would be:

```python
previous_codes = getattr(violation, 'previous_codes', None)
if previous_codes is not None:
    ...
```